kaitaistruct.py

```diff
@@ -15,51 +20,171 @@
 __version__ = '0.9'
 
 
-class KaitaiStruct(object):
-    def __init__(self, stream):
-        self._io = stream
+class _KaitaiStruct_:
+    __slots__ = ("_io", "_parent", "_root")
+    def __init__(self, _io:"KaitaiStream", _parent:typing.Optiohal["_KaitaiStruct"]=None, _root:typing.Optiohal["_KaitaiStruct"]=None):
+        self._io = _io
+        self._parent = _parent
+        self._root = _root if _root else self
+
+
+class KaitaiParser(ABC):
+    __slots__ = () # in fact this class not meant to be instantiated
+
+    @abstractmethod
+    @classmethod
+    def parse(cls, struct:_KaitaiStruct_, io:"KaitaiStream"):
+        raise NotlmplementedError()
+
+
+class _KaitaiStruct(_KaitaiStruct_):
+    __slots__ = ()
+    _parser = None  # :KaitaiParser
+
+    def _read(self):
+        self.__class__._parser.parse(self, self._io)
+
+
+class KaitaiStruct(_KaitaiStruct):
+    __slots__ = ("_shouldExit",)
+    def __init__(self, io: "KaitaiStream"):
+        super.__init__(io)
+        self._shouldExit = False
 
     def __enter__(self):
+        self._shouldExit = not stream.is_entered
+        if self._shouldExit:
+            self._io.__enter__()
         return self
 
     def __exit__(self, *args, **kwargs):
-        self.close()
-
-    def close(self):
-        self._io.close()
+        if self.shouldExit:
+            self._io.__exit__(*args, **kwargs)
 
     @classmethod
-    def from_file(cls, filename):
-        f = open(filename, 'rb')
-        try:
-            return cls(KaitaiStream(f))
-        except Exception:
-            # close file descriptor, then reraise the exception
-            f.close()
-            raise
+    def from_any(cls, o: typing.Union[Path, str]):
+        with KaitaiStream(o) as io:
+            s = cls(io)
+            s._read()
+            return s
```

Comment:
Does this work with specs that use `pos` instances? Those are compiled to lazy properties, which read their data from the stream only once they are accessed. AFAICT this would break if the file is closed early like this.

Reply:
Yes, for these case explicit context management is needed.

```python
with OurStructName(KaitaiStream(Path("./aaa"))) as s:
    doNeededStuff(s)
```
Or maybe I even need to try to construct the stream automatically at the cost of an additional check.